extensions/wikia/RobotsTxt/PathBuilder.class.php

```diff
@@ -0,0 +1,172 @@
+	 *
+	 * @param string $pageName the name of the page, ex. "Spezial:Alle_Seiten"
+	 * @param bool $canonicalOnly only return the canonical way of accessing the page/namespace
+	 * @return array an array of paths
+	 */
+	public function buildPaths( $pageName, $canonicalOnly = false ) {
+		$pageName = $this->encodeUrl( $pageName );
+		$paths = [
+			str_replace( '$1', $pageName, $this->articlePath )
+		];
+
+		if ( $canonicalOnly ) {
+			return $paths;
+		}
+
+		$paths[] = '/*?*title=' . $pageName;
+		$paths[] = '/index.php/' . $pageName;
+
+		return $paths;
+	}
+
+	/**
+	 * Get paths a given special page is accessible at
+	 *
+	 * @param string $pageName name of the special page as exposed in alias file for the special page
+	 * @param bool $canonicalOnly only return the canonical way of accessing the page
+	 * @return array of paths
+	 */
+	public function buildPathsForSpecialPage( $pageName, $canonicalOnly = false ) {
+		$paths = [];
+		foreach ( $this->specialNamespaces as $specialNamespaceAlias ) {
+			foreach ( $this->getSpecialPageNames( $pageName ) as $localPageName ) {
+				$paths = array_merge(
+					$paths,
+					$this->buildPaths( $specialNamespaceAlias . ':' . $localPageName, $canonicalOnly )
+				);
+			}
+		}
+		return $paths;
+	}
+
+	/**
+	 * Get paths a given namespace is accessible at
+	 *
+	 * @param int $namespaceId namespace ID
+	 * @param bool $canonicalOnly only return the canonical way of accessing the namespace
+	 * @return array of paths
+	 */
+	public function buildPathsForNamespace( $namespaceId, $canonicalOnly = false ) {
+		$paths = [];
+
+		foreach ( $this->getNamespaces( $namespaceId ) as $namespace ) {
+			$paths = array_merge(
+				$paths,
+				$this->buildPaths( $namespace . ':', $canonicalOnly )
+			);
+		}
+
+		return $paths;
+	}
+
+	/**
+	 * Build path matching URLs with a specific param present
+	 *
+	 * This will only work for robots that understand wildcards.
+	 *
+	 * @param string $param URL param to block
+	 * @return array
+	 */
+	public function buildPathsForParam( $param ) {
+		return [
+			'/*?' . $param . '=',
+			'/*?*&' . $param . '=',
+		];
+	}
+
+	/**
+	 * Encode URL in a way you can safely put that to robots.txt
+	 *
+	 * There's no need to encode characters like /, :, *, ?, &, =, $, so they are decoded
+	 * Non-English characters WILL be encoded.
+	 *
+	 * @param string $in the URL to encode
+	 * @return string
+	 */
+	private function encodeUrl( $in ) {
+		return str_replace(
+			[ '%2F', '%3A', '%2A', '%3F', '%26', '%3D', '%24' ],
+			[ '/', ':', '*', '?', '&', '=', '$' ],
+			rawurlencode( $in )
+		);
+	}
+
+	/**
+	 * Get the special page's aliases
+	 *
+	 * @param string $pageName special page name as specified in the SpecialPage object
+	 * @return array
+	 */
+	private function getSpecialPageNames( $pageName ) {
+		$aliases = [];
+		foreach ( $this->languages as $lang ) {
+			$specialAliases = $lang->getSpecialPageAliases();
+			if ( isset( $specialAliases[$pageName] ) ) {
```

Comment:
There's a hook that allows 3rd party extensions to break the structure, I'm adding an additional is_array here to be safe.